File: ADT/trees/94.py

class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        result = []
        curr = root
        while curr is not None:
            if curr.left is None:
                result.append(curr.val)
                curr = curr.right
            else:
                leftChild = curr.left
                while leftChild.right is not None:
                    leftChild = leftChild.right
                leftChild.right = curr
                temp = curr
                curr = curr.left
                temp.left = None
        return result

File: ADT/trees/94.cpp
// Morris Traversal

// Tc = O(n)
// Sc = O(1) [not used recursion so system stack is not used but with recursion for system stack depth]

class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> result;
        TreeNode* curr = root;
        while (curr != NULL) {
            if (curr->left == NULL) {
                result.push_back(curr->val);
                curr = curr->right;
            } else {
                TreeNode* leftChild = curr->left;
                
                while (leftChild->right != NULL) {
                    leftChild = leftChild->right;
                }
                
                leftChild->right = curr;
                TreeNode* temp = curr;
                curr = curr->left;
                temp->left = NULL;
            }
        }

        return result;
    }
};